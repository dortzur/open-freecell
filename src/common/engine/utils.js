import _ from 'lodash/fp';

export const getTopCard = (resourceTarget) => {
  if (resourceTarget.value) {
    return resourceTarget.value[resourceTarget.value.length - 1];
  }
  return resourceTarget[resourceTarget.length - 1];
};
export const getSuitFoundation = (state, suit) =>
  state.foundation.reduce((acc, foundationCell) => {
    if (!_.isEmpty(foundationCell) && foundationCell[0].suit === suit) {
      acc = foundationCell;
    } else if (!acc && _.isEmpty(foundationCell)) {
      acc = foundationCell;
    }
    return acc;
  }, null);

export const areSameColor = (cardA, cardB) => cardA.color === cardB.color;

export const isIncrementalValueDiff = (sourceCard, targetCard) =>
  getValueDiff(sourceCard, targetCard) === 1;

export const getValueDiff = (sourceCard, targetCard) =>
  sourceCard.value - targetCard.value;

export const isFoundationStackable = (sourceCell, foundationCell) => {
  const sourceCard = getTopCard(sourceCell);
  const foundationCard = getTopCard(foundationCell);

  return (
    areSameColor(sourceCard, foundationCard) &&
    isIncrementalValueDiff(sourceCard, foundationCard)
  );
};
