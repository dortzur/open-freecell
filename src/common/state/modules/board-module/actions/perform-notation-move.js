import * as types from '../types';
export const performNotationMove = (notation) => ({
  type: types.PERFORM_NOTATION_MOVE,
  payload: { notation },
});
