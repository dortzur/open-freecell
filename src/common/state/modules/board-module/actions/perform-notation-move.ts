import * as types from '../types';

export const performNotationMove = (notation: string) => ({
  type: types.PERFORM_NOTATION_MOVE,
  payload: { notation },
});
