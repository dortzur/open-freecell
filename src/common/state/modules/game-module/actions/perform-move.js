import * as types from '../types';
export const performMove = (notation) => ({
  type: types.PERFORM_MOVE,
  payload: { notation },
});
