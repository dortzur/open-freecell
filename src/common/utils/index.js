export const srand = (seed) => () => {
  seed = (seed * 214013 + 2531011) & 0x7fffffff;
  return seed >> 16;
};
