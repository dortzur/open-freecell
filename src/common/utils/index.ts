import { deck } from './consts';

export const srand = (seed: number) => () => {
  seed = (seed * 214013 + 2531011) & 0x7fffffff;
  return seed >> 16;
};

export const createGameDeck = (gameNumber: number) => {
  const gameDeck = [...deck];
  const rand = srand(gameNumber);
  for (let i = gameDeck.length - 1; i > 0; i--) {
    const r = rand() % (i + 1);
    const card = gameDeck[r];
    gameDeck[r] = gameDeck[i];
    gameDeck[i] = card;
  }
  gameDeck.reverse();
  return gameDeck;
};
