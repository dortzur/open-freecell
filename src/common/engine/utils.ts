import _ from 'lodash/fp';
import { Card, Cell } from '../utils/consts';
import { BoardState } from '../state/modules/board-module';


export const getTopCard = (resourceTarget: Cell): Card =>
  resourceTarget[resourceTarget.length - 1];

export const getBottomCard = (resourceTarget: Cell) => resourceTarget[0];

export const getSuitFoundation = (state: BoardState, suit: string): Cell =>
  <Cell>state.foundation.reduce((acc: null | Cell, foundationCell): Cell => {
    if (!_.isEmpty(foundationCell) && foundationCell[0].suit === suit) {
      acc = foundationCell;
    } else if (!acc && _.isEmpty(foundationCell)) {
      acc = foundationCell;
    }
    return <Cell>acc;
  }, null);

export const areSameColor = (cardA: Card, cardB: Card) => cardA.color === cardB.color;

export const isIncrementalValueDiff = (sourceCard: Card, targetCard: Card) =>
  getValueDiff(sourceCard, targetCard) === 1;

export const getValueDiff = (sourceCard: Card, targetCard: Card) =>
  sourceCard.value - targetCard.value;

export const isFoundationStackable = (sourceCell: Cell, foundationCell: Cell) => {
  const sourceCard = getTopCard(sourceCell);
  const foundationCard = getTopCard(foundationCell);

  return (
    areSameColor(sourceCard, foundationCard) &&
    isIncrementalValueDiff(sourceCard, foundationCard)
  );
};
