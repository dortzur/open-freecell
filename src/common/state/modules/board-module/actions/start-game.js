import * as types from '../types';
import { createGameDeck } from '../../../../utils';

export const startGame = (gameNumber) => {
  const gameDeck = createGameDeck(gameNumber);
  return {
    type: types.START_GAME,
    payload: { gameDeck, gameNumber },
  };
};
