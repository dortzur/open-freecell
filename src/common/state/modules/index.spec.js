import gameReducer from './game-module';
import { startGame } from './game-module/actions/start-game';
import { performNotationMove } from './game-module/actions/perform-notation-move';
import { print } from '../../utils/print';
describe('game-module', () => {
  let state = null;
  beforeEach(() => {
    state = gameReducer(undefined, startGame(100));
  });

  it('should perform legal cell moves', () => {
    state = gameReducer(state, performNotationMove('1a'));
    expect(state.tableau[0]).toHaveLength(6);
    expect(state.cell[0]).toHaveLength(1);

    state = expect(() =>
      gameReducer(state, performNotationMove('1a'))
    ).toThrow();
  });

  it('should perform legal foundation moves', () => {
    state = gameReducer(undefined, startGame(106));
    state = gameReducer(state, performNotationMove('2h'));
    state = gameReducer(state, performNotationMove('8h'));

    state = gameReducer(state, performNotationMove('7a'));
    state = gameReducer(state, performNotationMove('7b'));
    state = gameReducer(state, performNotationMove('7c'));
    state = gameReducer(state, performNotationMove('7d'));
    state = gameReducer(state, performNotationMove('7h'));
    expect(state.tableau[0]).toHaveLength(7);
    expect(state.foundation[0][0].id).toEqual('AC');
    expect(state.foundation[1][0].id).toEqual('AH');
    expect(state.foundation[1][1].id).toEqual('2H');
    expect(() => gameReducer(state, performNotationMove('2h'))).toThrow();
    expect(() => gameReducer(state, performNotationMove('7h'))).toThrow();
  });

  it('should perform legal tableau moves', () => {
    state = gameReducer(state, performNotationMove('63'));
    state = gameReducer(state, performNotationMove('8a'));
    state = gameReducer(state, performNotationMove('86'));
    state = gameReducer(state, performNotationMove('72'));

    expect(() => gameReducer(state, performNotationMove('63'))).toThrow();
    state = gameReducer(state, performNotationMove('81'));

    state = gameReducer(state, performNotationMove('1b'));
    state = gameReducer(state, performNotationMove('2c'));
    state = gameReducer(state, performNotationMove('2d'));
    expect(() => gameReducer(state, performNotationMove('67'))).toThrow();
    state = gameReducer(state, performNotationMove('b1'));
    state = gameReducer(state, performNotationMove('67'));
    expect(state.tableau[6][6].id).toEqual('3H');
    expect(state.tableau[5][3].id).toEqual('2S');
    expect(state.tableau[0][7].id).toEqual('7H');
    expect(state.tableau[3][6].id).toEqual('5S');
  });
  it('should perform auto moves', () => {
    state = gameReducer(state, performNotationMove('63'));
    state = gameReducer(state, performNotationMove('8a'));
    state = gameReducer(state, performNotationMove('86'));
    state = gameReducer(state, performNotationMove('72'));
    state = gameReducer(state, performNotationMove('81'));
    // state = gameReducer(state, performNotationMove('81'));


    print(state);
  });
});
