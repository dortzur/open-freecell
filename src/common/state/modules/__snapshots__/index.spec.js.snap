// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game-module should create game 100 1`] = `
Object {
  "cell": Array [
    Array [],
    Array [],
    Array [],
    Array [],
  ],
  "foundation": Array [
    Array [],
    Array [],
    Array [],
    Array [],
  ],
  "tableau": Array [
    Array [
      Object {
        "color": "red",
        "id": "AD",
        "rank": "A",
        "suit": "D",
        "value": 1,
      },
      Object {
        "color": "red",
        "id": "QH",
        "rank": "Q",
        "suit": "H",
        "value": 12,
      },
      Object {
        "color": "black",
        "id": "QS",
        "rank": "Q",
        "suit": "S",
        "value": 12,
      },
      Object {
        "color": "black",
        "id": "6S",
        "rank": "6",
        "suit": "S",
        "value": 6,
      },
      Object {
        "color": "black",
        "id": "3C",
        "rank": "3",
        "suit": "C",
        "value": 3,
      },
      Object {
        "color": "black",
        "id": "2C",
        "rank": "2",
        "suit": "C",
        "value": 2,
      },
      Object {
        "color": "black",
        "id": "8S",
        "rank": "8",
        "suit": "S",
        "value": 8,
      },
    ],
    Array [
      Object {
        "color": "black",
        "id": "JS",
        "rank": "J",
        "suit": "S",
        "value": 11,
      },
      Object {
        "color": "black",
        "id": "TS",
        "rank": "T",
        "suit": "S",
        "value": 10,
      },
      Object {
        "color": "red",
        "id": "TH",
        "rank": "T",
        "suit": "H",
        "value": 10,
      },
      Object {
        "color": "red",
        "id": "5H",
        "rank": "5",
        "suit": "H",
        "value": 5,
      },
      Object {
        "color": "black",
        "id": "KS",
        "rank": "K",
        "suit": "S",
        "value": 13,
      },
      Object {
        "color": "red",
        "id": "2H",
        "rank": "2",
        "suit": "H",
        "value": 2,
      },
      Object {
        "color": "black",
        "id": "KC",
        "rank": "K",
        "suit": "C",
        "value": 13,
      },
    ],
    Array [
      Object {
        "color": "black",
        "id": "4S",
        "rank": "4",
        "suit": "S",
        "value": 4,
      },
      Object {
        "color": "black",
        "id": "5C",
        "rank": "5",
        "suit": "C",
        "value": 5,
      },
      Object {
        "color": "black",
        "id": "7S",
        "rank": "7",
        "suit": "S",
        "value": 7,
      },
      Object {
        "color": "black",
        "id": "6C",
        "rank": "6",
        "suit": "C",
        "value": 6,
      },
      Object {
        "color": "black",
        "id": "JC",
        "rank": "J",
        "suit": "C",
        "value": 11,
      },
      Object {
        "color": "red",
        "id": "KH",
        "rank": "K",
        "suit": "H",
        "value": 13,
      },
      Object {
        "color": "black",
        "id": "9S",
        "rank": "9",
        "suit": "S",
        "value": 9,
      },
    ],
    Array [
      Object {
        "color": "black",
        "id": "QC",
        "rank": "Q",
        "suit": "C",
        "value": 12,
      },
      Object {
        "color": "red",
        "id": "9D",
        "rank": "9",
        "suit": "D",
        "value": 9,
      },
      Object {
        "color": "red",
        "id": "8H",
        "rank": "8",
        "suit": "H",
        "value": 8,
      },
      Object {
        "color": "red",
        "id": "3D",
        "rank": "3",
        "suit": "D",
        "value": 3,
      },
      Object {
        "color": "black",
        "id": "3S",
        "rank": "3",
        "suit": "S",
        "value": 3,
      },
      Object {
        "color": "black",
        "id": "8C",
        "rank": "8",
        "suit": "C",
        "value": 8,
      },
      Object {
        "color": "black",
        "id": "5S",
        "rank": "5",
        "suit": "S",
        "value": 5,
      },
    ],
    Array [
      Object {
        "color": "black",
        "id": "7C",
        "rank": "7",
        "suit": "C",
        "value": 7,
      },
      Object {
        "color": "red",
        "id": "4H",
        "rank": "4",
        "suit": "H",
        "value": 4,
      },
      Object {
        "color": "black",
        "id": "9C",
        "rank": "9",
        "suit": "C",
        "value": 9,
      },
      Object {
        "color": "red",
        "id": "4D",
        "rank": "4",
        "suit": "D",
        "value": 4,
      },
      Object {
        "color": "black",
        "id": "TC",
        "rank": "T",
        "suit": "C",
        "value": 10,
      },
      Object {
        "color": "red",
        "id": "7D",
        "rank": "7",
        "suit": "D",
        "value": 7,
      },
    ],
    Array [
      Object {
        "color": "red",
        "id": "6D",
        "rank": "6",
        "suit": "D",
        "value": 6,
      },
      Object {
        "color": "red",
        "id": "JD",
        "rank": "J",
        "suit": "D",
        "value": 11,
      },
      Object {
        "color": "red",
        "id": "6H",
        "rank": "6",
        "suit": "H",
        "value": 6,
      },
      Object {
        "color": "black",
        "id": "2S",
        "rank": "2",
        "suit": "S",
        "value": 2,
      },
      Object {
        "color": "black",
        "id": "4C",
        "rank": "4",
        "suit": "C",
        "value": 4,
      },
      Object {
        "color": "red",
        "id": "8D",
        "rank": "8",
        "suit": "D",
        "value": 8,
      },
    ],
    Array [
      Object {
        "color": "red",
        "id": "JH",
        "rank": "J",
        "suit": "H",
        "value": 11,
      },
      Object {
        "color": "red",
        "id": "KD",
        "rank": "K",
        "suit": "D",
        "value": 13,
      },
      Object {
        "color": "black",
        "id": "AC",
        "rank": "A",
        "suit": "C",
        "value": 1,
      },
      Object {
        "color": "red",
        "id": "9H",
        "rank": "9",
        "suit": "H",
        "value": 9,
      },
      Object {
        "color": "red",
        "id": "5D",
        "rank": "5",
        "suit": "D",
        "value": 5,
      },
      Object {
        "color": "red",
        "id": "QD",
        "rank": "Q",
        "suit": "D",
        "value": 12,
      },
    ],
    Array [
      Object {
        "color": "red",
        "id": "TD",
        "rank": "T",
        "suit": "D",
        "value": 10,
      },
      Object {
        "color": "red",
        "id": "AH",
        "rank": "A",
        "suit": "H",
        "value": 1,
      },
      Object {
        "color": "black",
        "id": "AS",
        "rank": "A",
        "suit": "S",
        "value": 1,
      },
      Object {
        "color": "red",
        "id": "7H",
        "rank": "7",
        "suit": "H",
        "value": 7,
      },
      Object {
        "color": "red",
        "id": "3H",
        "rank": "3",
        "suit": "H",
        "value": 3,
      },
      Object {
        "color": "red",
        "id": "2D",
        "rank": "2",
        "suit": "D",
        "value": 2,
      },
    ],
  ],
}
`;
